homeassistant/components/hunterdouglas_powerview/cover.py

```diff
@@ -277,9 +338,117 @@ async def async_added_to_hass(self):
     @callback
     def _async_update_shade_from_group(self):
         """Update with new data from the coordinator."""
-        if self._scheduled_transition_update:
-            # If a transition in in progress
-            # the data will be wrong
+        if self._scheduled_transition_update or self.coordinator.data is None:
+            # If a transition is in progress the data will be wrong
+            # or empty data as result of 204/423 return
             return
         self._async_process_new_shade_data(self.coordinator.data[self._shade.id])
         self.async_write_ha_state()
+
+
+class PowerViewTDBUShade(PowerViewShade):
```

Comment:
Please split this so you don't have the same class for top and bottom shades and can get rid of the branching with `if self._motor ==`

Reply:
I'm a little confused here sorry on what you want
PowerViewTDBUShade controls a single blind that has two motors - one that controls the top and one that controls the bottom 
Both of these motors have 90% of the same code exept for the if self._motor parts which work out which motor we want to use

All other shades would use the base PowerViewShade class

Are you saying you want to duplicate the class for PowerViewTDBUShade into say PowerViewTDBUShadeTOPSHADE and PowerViewTDBUShadeBOTTOMSHADE ?

Just feel this will be a lot of duplication of code.

The bottom shade needs the same checks the top one does to ensure it can know the top shades position - i thought the if self._motor checks in the new code were relatively minimal.

Defintely understood the wanting to split the class out with the first commit - just hoping to understand what we are trying to achieve duplicating the shadeclass

I would envisage longer term each shadeType would have its own class - not an individual class per motor

ultimately would be great if i didnt need seperate entities as i can post both positions at once and move to the correct position, currently i am limited to moving only one motor at a time because of the cover object limitations in HA

Picture for context 
![image](https://user-images.githubusercontent.com/7148670/155068572-d2fab5f0-7b31-43a9-a55d-ccc5f090ec52.png)

Reply:
Here is an incomplete psuedo example: https://github.com/bdraco/home-assistant/commit/470bb3613c275a4bec8c3f2e23518970176f678f

Reply:
thanks - i can work on splitting that if thats how we want it - but the blind still needs to know the position of the other motor too to stop it jamming in the middle (that was a fun experience to deal with)

Ideally we could extend the Cover class for a "Dual Cover" and then the interface could handle it like this mockup (can move both sides) and the shade would be a single entity in HA (how it should be controlled)

Any tips or opinion on this?


![image](https://user-images.githubusercontent.com/7148670/155257289-99476e0c-a3d1-47b3-90d8-bfb5eec66fe3.png)

Reply:
Hi @bdraco - get this would be a bit more complicated than extending the powerview integration but what are your thoughts here?

Reply:
For that to be possible we would need to change the entity model which requires an arch discussion here https://github.com/home-assistant/architecture/discussions

Reply:
@bdraco  could you give me a hand with this one - i was trying to implement the changes suggested in the architecture forum and including your suggestion from here https://github.com/bdraco/home-assistant/commit/470bb3613c275a4bec8c3f2e23518970176f678f

issue is mypy wont validate it (the code works but mypy fails it)
any suggestions ?
`homeassistant/components/hunterdouglas_powerview/cover.py:113: error: Argument 1 to "append" of "list" has incompatible type "Type[PowerViewShade]"; expected "Union[PowerViewTDBUShadeTop, PowerViewTDBUShadeBottom, PowerViewShade]"  [arg-type]
homeassistant/components/hunterdouglas_powervie`
I was also implementing tilt support with some feedback from the community

Reply:
I think its going to get a bit easier once we merge https://github.com/home-assistant/core/pull/70775 as it creates the base class you need.

Reply:
mmm i was using the standard shades as the 'base' class - not sure that is the issue here. i saw no need for a new 'base' class as other blinds extend the usage of standard shades anyway

interesting pull though - glad some of the code i shared was perhaps useful, shame i didnt get a commit in first for it

i would maybe suggest for that pull that the Silhoutte not extend the base shade but rather their be another class thay extends the base then the silhoutte extends it - if you check the forums this is 1 of 4 shades and the other 3 function alike (i was trying to cater for them all)

the tilt functionality he has implemented is a niche blind - the standard HD tilt wont work with his code as they have a different max position - i would suggest setting the base class afound that then extending for the niche not the other way.

Reply:
Moved to other class type